src/FSharp.Data.GraphQL.Shared/Validation.fs

```diff
@@ -1,60 +1,1057 @@
+        function
+        | Field field -> 
+            let path = field.AliasOrName :: path
+            match field.SelectionSet |> List.map (checkFragmentTypeExistenceInSelection acc fragmentDefinitions schemaInfo path) with
+            | [] -> acc
+            | results -> List.reduce (@) results
+        | InlineFragment frag -> checkFragmentTypeExistence acc fragmentDefinitions schemaInfo path frag
+        | _ -> acc
+       
+    let internal validateFragmentTypeExistence (ctx : ValidationContext) =
+        let fragmentDefinitions = getFragmentDefinitions ctx.Document
+        ctx.Document.Definitions
+        |> List.fold (fun acc def ->
+            match def with
+            | FragmentDefinition frag -> 
+                let path = match frag.Name with | Some name -> [name] | None -> []
+                checkFragmentTypeExistence acc fragmentDefinitions ctx.Schema path frag
+            | OperationDefinition odef -> 
+                let path = match odef.Name with | Some name -> [name] | None -> []
+                match odef.SelectionSet |> List.map (checkFragmentTypeExistenceInSelection acc fragmentDefinitions ctx.Schema path) with
+                | [] -> acc
+                | results -> List.reduce (@) results) Success
+
+    let rec private checkFragmentOnCompositeType (acc : ValidationResult) (selection : SelectionInfo) =
+        let acc =
+            match selection.FragmentType with
+            | Some fragType ->
+                match fragType.Kind with
+                | TypeKind.UNION | TypeKind.OBJECT | TypeKind.INTERFACE -> acc
+                | _ when selection.FragmentSpreadName.IsSome -> acc @ Error.AsResult(sprintf "Fragment '%s' has type kind %s, but fragments can only be defined in UNION, OBJECT or INTERFACE types." selection.FragmentSpreadName.Value (fragType.Kind.ToString()), selection.Path)
+                | _ -> acc @ Error.AsResult(sprintf "Inline fragment has type kind %s, but fragments can only be defined in UNION, OBJECT or INTERFACE types." (fragType.Kind.ToString()), selection.Path)
+            | None -> acc
+        selection.SelectionSet |> List.fold (fun acc selection -> checkFragmentOnCompositeType acc selection) acc
+
+    let internal validateFragmentsOnCompositeTypes (ctx : ValidationContext) =
+        ctx.Definitions
+        |> List.collect (fun def -> def.SelectionSet)
+        |> List.fold (fun acc selection -> checkFragmentOnCompositeType acc selection) Success
+
+    let internal validateFragmentsMustBeUsed (ctx : ValidationContext) =
+        let rec getSpreadNames (acc : string list) =
+            function
+            | Field field -> field.SelectionSet |> List.collect (getSpreadNames acc)
+            | InlineFragment frag -> frag.SelectionSet |> List.collect (getSpreadNames acc)
+            | FragmentSpread spread -> spread.Name :: acc
```

Comment:
Instead of using a list and then doing `List.distinct`, perhaps we should use a set?

Reply:
Did a change to use sets, not sure if it's the best, please take a look.